### ISSUE TITLE
Lightweight Statistics - Research 

### ISSUE THREAD

**Issue by bosmatt** [speedb-io/speedb#255]
As part of the lightweight statistics feature we would like to test the following:
1. Measure the amount of system resources being consumed by the existing statistics mechanism.
2. POC version that collects statistics every X operations instead of every operation. Find an X that provides a balance between useful statistics information and system performance.
3.  If the system resources are dramatically reduced with this method, continue developing the full feature.

**Comment by udi-speedb** [comment]
@RoyBenMoshe Please also measure the performance using the existing statistics level that bypasses the timers. This is to verify that indeed the problem is time measurements. Thanks

**Comment by RoyBenMoshe** [comment]
Here is the results and conclusions of a research we have done on the performance degradation caused by turning on the statistics.
We ran thousands of iterations on many scenarios and configurations in order to find the most "affected by statistics" scenario and measured the performance degradation.

1. The results shows that the most statistics affected scenario is read only scenario small block on a small database, the performance degradation was measured was 10% on average.
Below is a graph and histograms showing some of our results:

![Image](https://user-images.githubusercontent.com/112958533/216417225-385e9f20-2ba4-48c2-8f29-be1315c59aa6.png)



![Image](https://user-images.githubusercontent.com/112958533/216417509-cda1ffe4-e06e-4e97-b4c9-6d71abae260c.png)

**We found this number consistent in this scenario for many configurations and we haven’t found scenario and configuration that shows higher degradation.**

2. In order to check the time measurement influence on the performance degradation we have created POC version that not measuring time at all. The results on this branch for the same scenario and configuration showed performance degradation of about 7% on average
 

![Image](https://user-images.githubusercontent.com/112958533/216417774-a5829381-0e60-40fa-b2af-6a9cd7861f28.png)



![Image](https://user-images.githubusercontent.com/112958533/216417862-55cd2993-ab8d-4e50-81ce-87764467e944.png)


3. According to these results we believe that the drop in performance is **no more than 10%** on average due to running statistics and the improvement in statistical time measurement **will not make it better than 7%**.

According to these numbers, we don't think that the system resources are being reduced dramatically, so we decided not to continue developing the full feature.

**Comment by mrambacher** [comment]
Can you give more information about the tests you executed and how you reached these conclusions?  
- Is most of the time in collecting individual stats (e.g. retrieving time and calculating durations/buckets) or collating statistics between threads (mutex/locks)
- What tests were you executing when you performed these tests?
- Unless I am reading the code wrong, it looks as if there are 5 stats level (disable, except histograms, except timers, except detailed timers, except time for mutex, all).  
- When stats == 1 (I think this is kExcept timers), stats should be doing a simple counting.  Is there any idea why the overhead here is so high?
- Is there any effort to say where the stats are spending their time during collection (e.g. compare a flame graph of runs at various stats levels to see where/why statistics are spending their time?
- How (if at all) does gathering statistics change at different levels affect the p99 performance?

**Comment by RoyBenMoshe** [comment]
The purpose of this study was to test the effect on performance of turning on statistics or more clearly-
what is the performance degradation between enabling statistics flag at default level or leaving it off,
We had an old assumption that it might be more than 30% and we wanted to test it.

I executed db_bench scenarios, together with our scientists and performance team, we tuned the scenario and the configuration till it showed the biggest difference.

This is the chosen scenario in short:
    1. fillrandom 
    2. waitforcompaction
    3. for i iterations
            * readrandom w statistics
            * readrandom w/o statistics 

* As a part of this research we did not determine if most of the time is for collecting individual stats or collecting statistics between threads that might be a good subject for further research.
* You are not wrong we didn't check the all level but as I wrote it was not in the scope. In most of the iterations we just executed default level statistics against without statics, but I thought that these results give some additional value so I decided to attach it.

Like these questions the others also were not in the scope of this research but may be subjects for further research.
